### PR TITLE
add pos_weight hyperparameter for dealing with imbalanced dataset #2703

### PR DESCRIPTION
**Pull request by qllesudo**
- Related issue: https://github.com/ultralytics/ultralytics/issues/2703#
- Description:
This PR add a new feature to Ultralytics YOLOv8 Detection by adding pos_weight to the [BCEWithLogitsLoss](https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html) for dealing with imbalanced dataset.
- Example:
```
model = YOLO("yolov8s.pt")
model.train(data="coco.yaml", pos_weight=[0.5, 2]) # pos_weight must be the same length as class dimension.
```
_I have read the CLA Document and I hereby sign the CLA_

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Addition of class imbalance handling with `pos_weight` in loss calculation.

### 📊 Key Changes
- Introduced a new configuration option `pos_weight` in the default settings (`default.yaml`).
- Updated the loss calculation logic in `loss.py` to consider the `pos_weight` for the binary cross-entropy calculation.

### 🎯 Purpose & Impact
- 🎯 **Purpose**: To improve model performance on imbalanced datasets by adjusting the weight of positive samples in loss computation.
- 👥 **Impact**: Users working with datasets that have class imbalances may notice improved model training results as the model now accounts for this imbalance, potentially leading to better generalization and performance.

### CODE REVIEW

**Comment by hulkds** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Burhan-Q** [comment]
@hulkds thank you for your work!

I don't have final say on if your changes will get integrated or not, but I'd like to provide some feedback for you to help improve the chances that it could get accepted. I have not tested the new code, these are just pointers from my review of the changes made.

1. You have included a new initialization for `self.bce` on L171 however the original on L156 remains, which means it would be initialized twice. It' probably better to move your new line up to replace the old one and insert the line for `self.pos_weight` above.
1. I see you added this new argument to the configuration YAML, but have you verified that it works? When you add the argument for training like you show in your example 

    ```py
    model = YOLO("yolov8s.pt")
    model.train(
        data="coco.yaml",
        pos_weight=[0.5, 2],  # pos_weight must be the same length as class dimension.
        )
    ```
    are you certain that this change is applied? If so, can you share your results that demonstrate the applied change?
    
1. Have you tested what implications there are when changing the `pos_weight` argument for a _balanced_ dataset?
1. Changes that will update arguments should also be incorporated into the documentation. Please be sure to add your changes to docs as well.

**Comment by satyrmipt** [comment]
@hulkds , please answer Burhan-Q's questions. We all are waiting for your improvement to be implemented. In particular those of us who use Google Colab right now have no workarounds except for over\undersampling. For his 3rd question i think it's user responsibility to use correct weights. Default weights must be ones anyway.

**Comment by hulkds** [comment]
@Burhan-Q thanks for your feedback!
1. This is my bad, I'll fix it :)
2. I confirmed that the pos_weight argument is functioning as expected. After printing `self.hyp`, I noticed the inclusion of `pos_weight`. Furthermore, I can access `self.hyp.pos_weight` without any issues, indicating that the addition of `pos_weight` to the configuration YAML file is indeed effective.
3. I successfully trained my model incorporating this modification and observed a significant improvement in detecting vehicles and license plates (with a car to license plate ratio in my dataset is nearly 6:1), thanks to the `pos_weight` parameter.
4. I plan to submit another pull request soon. This PR will address your initial suggestion and include updates to the documentation to reflect these changes.

**Comment by Burhan-Q** [comment]
@hulkds thanks for the follow up! I checked out this PR and was able to verify that it works as well for both the case when adding weights or using the default value. I tested with the `coco128` dataset and did not encounter any issues. I _did not_ test with any other task `segment`, `classify`, `obb`, or `pose` but since the CI tests are passing, it seems that it's unlikely to be an issue for those tasks. Obviously with a limited test like mine, it's difficult to observe the differences. Later I will test with your changes versus without your changes to ensure that for the default there is no change in outcomes.

It would be preferable if you could checkout this PR or push commits to [your fork](https://github.com/hulkds/ultralytics/tree/fix%232703) for the documentation updates so that everything is included with a single PR. This will help to ensure your PR has _the best chance_ to get accepted (I can't say for certain it will tho).

**Comment by hulkds** [comment]
@Burhan-Q 
The change was made specifically to the `v8DetectionLoss` class, so I don't expect it to affect other tasks, but it's still good to double-check. I've created a new branch and pushed a commit [there](https://github.com/hulkds/ultralytics/commit/77fb118eb83f3e95a8aef6a95d6b5f7116717914). If everything looks fine, I will then create a new PR like so, everything will be included with a single PR.

**Comment by Burhan-Q** [comment]
@hulkds yes your updates look like they should be okay. The reason I mentioned the other tasks is because they inherit from the `v8DetectionLoss` for example, the Segmentation Loss https://github.com/ultralytics/ultralytics/blob/906b8d31dc5592dcef47bda9fc8258985591787e/ultralytics/utils/loss.py#L250

I checked the results from the existing training (without `pos_weight`) code against the training code with `pos_weight=[1]` and the results from my limited test seem to match. Seems good to me overall. However you'd like to commit your changes, it's up to you, but if you make a PR it would be good to close this PR and mention it in the new one. 🚀

**Comment by hulkds** [comment]
@Burhan-Q 
Here is the PR: https://github.com/ultralytics/ultralytics/pull/8620

**Comment by ethyljoybadiang** [comment]
Hello. I am dealing with an imbalanced dataset. How can I apply this concept to training? I tried this and it did not work. Thank you.

> @hulkds thank you for your work!
> 
> I don't have final say on if your changes will get integrated or not, but I'd like to provide some feedback for you to help improve the chances that it could get accepted. I have not tested the new code, these are just pointers from my review of the changes made.
> 
> 1. You have included a new initialization for `self.bce` on L171 however the original on L156 remains, which means it would be initialized twice. It' probably better to move your new line up to replace the old one and insert the line for `self.pos_weight` above.
> 2. I see you added this new argument to the configuration YAML, but have you verified that it works? When you add the argument for training like you show in your example
>    ```python
>    model = YOLO("yolov8s.pt")
>    model.train(
>        data="coco.yaml",
>        pos_weight=[0.5, 2],  # pos_weight must be the same length as class dimension.
>        )
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    are you certain that this change is applied? If so, can you share your results that demonstrate the applied change?
> 3. Have you tested what implications there are when changing the `pos_weight` argument for a _balanced_ dataset?
> 4. Changes that will update arguments should also be incorporated into the documentation. Please be sure to add your changes to docs as well.
